### ISSUE TITLE
Unable to parse SQL (Dialects.TSQL)

### ISSUE THREAD

**Issue by Habeeb556** [apache/superset#31183]
### Bug description

After upgrade from `4.0.1` to `4.1.1` I encountered an issue with permission 

```
DB engine Error
You may have an error in your SQL statement. Unable to parse script
```

This error occurred for users to whom I assigned specific permissions on the `dbo` schema only. When I changed the permissions to full access for the entire database `id:`, the issue was resolved. I noticed that this problem only happens with the `mssql.py` engine, while the others (MySQL, PostgreSQL, and Oracle) are working fine.

Log:

`WARNING:superset.sql_parse:Unable to parse SQL (Dialects.TSQL): Template(body=[Output(nodes=[TemplateData(data='SELECT ...')])])`

### Screenshots/recordings

![image](https://github.com/user-attachments/assets/0a63a19b-9df5-4b23-b82d-646f9f7cdfa0)

### Superset version

master / latest-dev

### Python version

3.11

### Node version

I don't know

### Browser

Not applicable

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by Habeeb556** [comment]
@dosu How Can I handle this as before upgrade?

**Comment by Habeeb556** [comment]
I have workaround this, using the old `sql_parse.py` script.

**Comment by geido** [comment]
@betodealmeida this might also use your help! 🙏

**Comment by betodealmeida** [comment]
> This error occurred for users to whom I assigned specific permissions on the `dbo` schema only. When I changed the permissions to full access for the entire database `id:`, the issue was resolved. I noticed that this problem only happens with the `mssql.py` engine, while the others (MySQL, PostgreSQL, and Oracle) are working fine.

Right, we need to parse the query in order to determine which catalogs/schemas/tables are being accessed when the user doesn't have full access to the database. If the user has full access we just don't parse the query.

In the past this was done using `sqlparse`, which is very lenient, so it was much harder to get an error when parsing SQL. The problem was that it was much easier to trick the parser, and the parsing was not dialect-specific, so we had many security holes.

We're currently in the process of switching the parsing to `sqlglot`, which has dialect-specific parsing. Unfortunately some engines don't have corresponding `sqlglot` dialects, and we have to fallback to a generic parser.

For MSSQL we use the T-SQL dialect, which AFAIK is the correct one. If you can share the query that's causing problems (anonymized is fine) we can take a look, and file a ticket upstream with `sqlglot`.

**Comment by betodealmeida** [comment]
Assigned it to you temporarily, @Habeeb556, since we need the query to investigate this more.

**Comment by Habeeb556** [comment]
Hello @betodealmeida, actually this issue isn't limited to a specific query. If you attempt to connect any MSSQL database with a dummy table and select from it while granting permissions on a specific schema (set to the default `dbo`), you will encounter this error.

![image](https://github.com/user-attachments/assets/2182afcd-f8cc-4408-aacc-b94f5754bf91)

**Comment by betodealmeida** [comment]
@Habeeb556 even if it's not specific, can you share a query where that happens?

**Comment by betodealmeida** [comment]
Thanks! I'll take a look!

**Comment by betodealmeida** [comment]
OK, I was able to repro this, working on a fix.

**Comment by betodealmeida** [comment]
Fix is out, as a workaround you can enable the `ENABLE_TEMPLATE_PROCESSING` feature flag.

**Comment by Habeeb556** [comment]
Awesome! I tried the workaround by adding `ENABLE_TEMPLATE_PROCESSING = True` in the `superset_config.py`, but it didn't work. However, after applying the fix you suggested in `sql_parse.py`, it's working perfectly. I really appreciate your help!

**Comment by Habeeb556** [comment]
Awesome! I tried the workaround by adding `ENABLE_TEMPLATE_PROCESSING = True` in the `superset_config.py`, but it didn't work. However, after applying the fix you suggested in `sql_parse.py`, it's working perfectly. I really appreciate your help!

**Comment by Habeeb556** [comment]
@betodealmeida, I have another inquiry also regarding this. Previously `4.0.1` when I select for example `select * from database_name.schema_name.table_name` it was working fine.

I applied the fix but still cannot select with database_name access denied. I required to grant full access `id:` to do that which wasnt before.

![image](https://github.com/user-attachments/assets/e23f89f7-2e25-41ff-b030-5badeb6c0a47)

```
Dec  8 11:26:10 ef-extsuperset superset[3696623]: 2024-12-08 11:26:10,802:DEBUG:superset.sql_parse:Parsing with sqlparse statement: select * from abcm.dbo.temp ;
Dec  8 11:26:10 ef-extsuperset superset[3696623]: 2024-12-08 11:26:10,816:DEBUG:superset.stats_logger:#033[36m[stats_logger] (incr) SqlLabRestApi.execute_sql_query.error#033[0m
Dec  8 11:26:10 ef-extsuperset superset[3696623]: 2024-12-08 11:26:10,816:DEBUG:superset.stats_logger:#033[36m[stats_logger] (timing) SqlLabRestApi.execute_sql_query.time | 122.99154791980982 #033[0m
Dec  8 11:26:11 ef-extsuperset superset[3696616]: 2024-12-08 11:26:11,842:DEBUG:superset.stats_logger:#033[36m[stats_logger] (incr) log#033[0m
```
but I am able to select with dbo only:

![image](https://github.com/user-attachments/assets/41bafd0a-1208-4bc1-b15f-231c1f89c818)

This needed for databases I have in the same instance and calling each other.
